# `snake-EAGLE-RC`

[![Snakemake](https://img.shields.io/badge/snakemake-≥6.3.0-brightgreen.svg)](https://snakemake.github.io)


**/!\ NOTE: This tools is only available for Linux /!\\**

A Snakemake workflow around [EAGLE-RC](https://github.com/tony-kuo/eagle?tab=readme-ov-file#eagle-rc), a tool designed to assign sequencing reads to its most likely genome of origin given several possible reference genomes. This is very useful for assigning reads from an allopolyploid to a specific subgenome. 

This workflow (snake-EAGLE-RC) is made to facilitate the read sorting process. It does this by automating the whole process so that you can easily process a large number of samples. The workflow also aims to improve reproducibility by providing a reproducibility report at the end which contains the md5sums of each input and (some) output files and the version (or download date) of each tool used in the analysis. 

*snake-EAGLE-RC* works for RNA, DNA and WGBS data and allows any number of genomes (any ploidy level for allopolyploids). It performs only the most basic steps required for read sorting. Namely, it aligns the reads to each reference genome and passes these aligments to EAGLE-RC for sorting. Additionnaly, a few quality checks are performed on the input data and alignment. The are done with fastqc and qualimap and reported in 'results/MultiQC/multiqc_report.html'.

**IMPORTANT:** This means it assumes the raw data has already been processed appropriately. Depending on your data this could mean clipping adapters, removing low quality reads or checking bisulfite conversion rates. The tools also does not remove duplicate reads as this is desirable only in some situations. You can judge if you wish to do that once the reads have been sorted or remove them pre-alignment. If you wish to (double)check the quality of the data before running the whole workflow you can run it only until fastqc (see Analysis section below). 

## Installation 

- [install Snakemake via Conda](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html).
- git clone snake-EAGLE-RC: 
```
git clone https://github.com/kenji-yt/snake-EAGLE-RC.git
```

## Input 

The input to snake-EAGLE-RC is a directory. The name of the directory must be the data type contained inside. Allowed options are: DNA, RNA or WGBS. If your data comes from an rna-seq experiment your input folder must be called 'RNA'. Inside it you should have two directories names 'progenitors' and 'polyploids': 
    - The 'progenitors' directory should contain one subdirectory for each reference genome (each progenitor for allopolyploids). The name of each subdirectory should be the species or genome name and it will appear as such in the output files. Within each subdirectory you should have the corresponding reference genome in fasta/fastq format. Name these files as you wish as long as they have one of the following extensions: "gff","gff3","fa","fasta","fq","fna","fastq". 
    - The 'polyploids' directory should contain one subdirectory per sample. The name of each subdirectory should be a unique sample identifier (don't give the same name to different sample directories). Each of these sample directories should contain (appropriately filtered/pre-processed) read data in fasta/fastq format. There should be one file if the data is single-end and two files if paired-end. Any other number of files will produce an error. You can have single end data for some samples and paired end data for others. 


Your input directory should have the following structure:
```
input_directory/
├── progenitors/
│   ├── species_1
│   │   └── assembly.fa
│   ├── species_2
│   │   └── assembly.fa
│   └── species_3
│       └── assembly.fa
└── polyploids/
    ├── sample_1
    │   ├── reads_pe_1.fastq
    │   └── reads_pe_2.fastq
    ├── sample_2
    │   └── reads_se.fastq
    └── sample_3
        ├── reads_pe_1.fastq
        └── reads_pe_2.fastq
```


## Analysis 

That's it, you are ready to sort reads using EAGLE-RC. From within the "snake-EAGLE-RC/" directory run:
```
snakemake --use-conda --cores N --config INPUT='your/input/directory'
```
Make sure to have snakemake make installed, to replace 'your/input/directory' with the path to you input directory and 'N' with the number of cores you wish to allocate to the job. If you installed snakemake in a conda environment make sure to activate it (eg. "conda activate snakemake_env").  

The outputs will now be generated in a results directory within the snake-EAGLE-RC directory. 


#### Memory allocation
Some steps in the analysis will automatically determine how much memory they will request. You can set maximal resource allocation using the flag:
```
snakemake --resources mem_mb='max_memory' 
```
With 'max_memory' replaced by the maximal amount of memory you wish the run to use in megabyte. 
Make sure to also add the other flags shown above. 

#### Quick Quality Check
If you want to use the workflow to automatically verify the quality of the input data for all samples you can run only the fastqc step:
```
snakemake --rule fastqc 
```
Make sure to also add the other flags shown above. 
